src/main/java/ch/hevs/cloudio/endpoint/CloudioEndpoint.java

```diff
@@ -226,6 +227,103 @@ public CloudioEndpoint(String uuidOrAppName) throws InvalidUuidException, Invali
         internal = new InternalEndpoint(uuidOrAppName, null, null);
     }
 
+    //TODO WIP Here
+    public void addNodesFromJson(String jsonPath) throws CloudioFactoryException {
```

Comment:
How about to implement the core behaviour of this method with an InputStream as parameter and add a wrapper for the JSON file path. This way a developer could bootstrap the endpoint/the node from any other input stream sources like for example an URL or from inside the jar file.

Reply:
@cm0x4D my idea was to have to have a string as parameter that will be the input the ResourceLoader to get the factory configuration a bit the same as how the certificates are handled. 

Also I thought about adding a parameter to let the user give a factory input without having to code it but still let this method public for more advanced user what do you think?

Since I added the interface to deserialize the factory structure, it's no more relevent to call it addNodesFromJson(String jsonPath). What about addNodesFromFactory(String factoryURI)?

Reply:
I am not sure to understand. My idea would be something like this:

```java
public void addNodesFromStream(InputStream nodesInputStream, CloudioFactoryFormat factoryFormat) throws CloudioFactoryException {
        try {
            CloudioFactoryNodes cloudioFactoryNodes = factoryFormat.deserializeNodes(nodesInputStream);
            for (String nodeKey : cloudioFactoryNodes.nodes.keySet()) {
                CloudioFactoryNode cloudioFactoryNode = cloudioFactoryNodes.nodes.get(nodeKey);

                CloudioDynamicNode cloudioDynamicNode;

                if (cloudioFactoryNode.type.equals("CloudioNode")) {
                    cloudioDynamicNode = new CloudioDynamicNode();
                    this.parseCloudioFactoryNode(cloudioFactoryNode, cloudioDynamicNode);
                    System.out.println(nodeKey);

                    this.addNode(nodeKey, cloudioDynamicNode);
                } else {
                    Class<?> nodeClass = Class.forName(cloudioFactoryNode.type);
                    Constructor<?> constructor = nodeClass.getConstructor();
                    Object nodeClassInstance = constructor.newInstance();

                    this.parseCloudioFactoryNode(cloudioFactoryNode, (CloudioDynamicNode) nodeClassInstance);

                    this.addNode(nodeKey, (CloudioNode) nodeClassInstance);
                }
            }
        } catch (Exception exception) {
            throw new CloudioFactoryException(exception);
        }
    }

    public void addNodesFromStream(InputStream nodesInputStream) throws CloudioFactoryException {
        addNodesFromStream(nodesInputStream, this.internal.factoryFormat);
    }

    public void addNodesFromResource(String path) throws CloudioFactoryException {
        try {
            InputStream jsonNodesInputStream = ResourceLoader.getResourceFromLocations(jsonPath,
                    this,
                    "home:" + "/.config/cloud.io/",
                    "file:/etc/cloud.io/",
                    "classpath:cloud.io/");
            addNodesFromStream(jsonNodesInputStream);
        }
        catch (FileNotFoundException exception) {
            throw new CloudioFactoryException("Json file "+jsonPath+" given to add Nodes From Json was not found " +
                    "[\"home:/.config/cloud.io/" + jsonPath + "\", " +
                    "\"file:/etc/cloud.io/" + jsonPath + "\", " +
                    "\"classpath:" + jsonPath + "\"].");
        } catch (Exception exception) {
            throw new CloudioFactoryException(exception);
        }
    }
```

The first method is fully custom, a developer can provide the input stream from where to read the definition and the factory that deserializes it. The second one takes just the input stream and uses the default deserializer by calling the 1st version. The third version creates an input stream from the resource locations and calls the 2nd method.

Reply:
Ooooh alright that's really nice like that, give more possibilities!